### PR TITLE
Make a better build scheme

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Build/Testing/Packaging Improvement

Short description (up to few sentences):

### CODE REVIEW

File: cmake/find_cxx.cmake [hunk 1]
```diff
@@ -1,26 +1,44 @@
...
+        find_library (LIBCXXFS_LIBRARY c++fs)
+        find_library (LIBCXXABI_LIBRARY c++abi)
+    else ()
+        set (LIBCXX_LIBRARY cxx)
+        set (LIBCXXABI_LIBRARY cxxabi)
+        add_subdirectory(contrib/libcxxabi-cmake)
+        add_subdirectory(contrib/libcxx-cmake)
+    endif ()
+
+    link_libraries(${LIBCXX_LIBRARY} ${LIBCXXABI_LIBRARY} ${LIBCXXFS_LIBRARY})
```
Comment (alexey-milovidov) [inline cmake/find_cxx.cmake]: Make sure that it will place libraries in appropriate order.

File: cmake/find_unwind.cmake [hunk 2]
```diff
@@ -1,59 +1,32 @@
...
-    set (USE_INTERNAL_UNWIND_LIBRARY 1)
-
-    set (PACKAGE_VERSION "9.0.0svn" CACHE STRING "")
-
-    set (UNWIND_INCLUDE_DIR "${ClickHouse_SOURCE_DIR}/contrib/libunwind/include")
-
-    set (LIBUNWIND_ENABLE_SHARED OFF CACHE BOOL "")
-    set (LIBUNWIND_ENABLE_STATIC ON CACHE BOOL "")
-    set (UNWIND_LIBRARY unwind_static)
+    # If we don't use built-in libc++abi, then we have to link directly with an exception handling library
```
Comment (alexey-milovidov) [inline cmake/find_unwind.cmake]: And if we do?

File: contrib/libcxxabi-cmake/CMakeLists.txt [hunk 3]
```diff
@@ -25,10 +22,14 @@ ${LIBCXXABI_SOURCE_DIR}/src/cxa_vector.cpp
...
-target_compile_definitions(cxxabi_static PRIVATE -D_LIBCPP_BUILDING_LIBRARY)
-target_compile_options(cxxabi_static PRIVATE -nostdinc++ -fno-sanitize=undefined) # If we don't disable UBSan, infinite recursion happens in dynamic_cast.
+add_library(cxxabi ${SRCS})
 
+target_include_directories(cxxabi PUBLIC ${LIBCXXABI_SOURCE_DIR}/include)
+target_compile_definitions(cxxabi PRIVATE -D_LIBCPP_BUILDING_LIBRARY)
+target_compile_options(cxxabi PRIVATE -nostdinc++ -fno-sanitize=undefined) # If we don't disable UBSan, infinite recursion happens in dynamic_cast.
 
+if (USE_INTERNAL_UNWIND_LIBRARY_FOR_EXCEPTION_HANDLING)
+    target_link_libraries(cxxabi PRIVATE ${UNWIND_LIBRARY})
```
Comment (alexey-milovidov) [inline contrib/libcxxabi-cmake/CMakeLists.txt]: Ok.

**Comment by alexey-milovidov** [comment]
FYI. When I tried to use `link_libraries`, CMake placed these libraries in two places: at the front of all other libraries and at the end. Example:

`glibc_compatibility.a libc.so other libraries glibc_compatibility.a libc.so`

And it doesn't work properly.

File: CMakeLists.txt [hunk 4]
```diff
@@ -421,19 +318,15 @@ if (UNBUNDLED)
...
 
-if (USE_LIBCXX)
-    set (HAVE_LIBCXX 1)
-    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libc++")
-endif()
-
-if (USE_LIBCXX AND USE_INTERNAL_LIBCXX_LIBRARY)
-    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -nostdinc++ -isystem ${LIBCXX_INCLUDE_DIR} -isystem ${LIBCXXABI_INCLUDE_DIR}")
-endif ()
+#if (USE_LIBCXX AND USE_INTERNAL_LIBCXX_LIBRARY)
```
Comment (abyss7) [inline CMakeLists.txt]: TODO: don't forget to uncomment this

**Comment by abyss7** [comment]
> FYI. When I tried to use `link_libraries`, CMake placed these libraries in two places: at the front of all other libraries and at the end. Example:
> 
> `glibc_compatibility.a libc.so other libraries glibc_compatibility.a libc.so`
> 
> And it doesn't work properly.

I hope I overcame it using INTERFACE libraries

**Pull request merged**